### PR TITLE
Extract Signup utils to fix crash

### LINKED ISSUE #12399: StringIndexOutOfBoundsException in SignupEpilogueFragment.createDisplayNameFromEmail
**Sentry Url**: https://sentry.io/share/issue/ad03935f8e3d4eb2a46e8168e3e3c54d/
**User Count**: 31
**Count**: 88
**First Release**: org.wordpress.android-12.3.1
**First Seen**: 2019-05-11T12:59:11Z
**Last Seen**: 2020-07-08T15:43:35Z
**24 Hours**: 4
**30 Days**: 23

```
StringIndexOutOfBoundsException: length=0; index=1
     at org.wordpress.android.ui.accounts.signup.SignupEpilogueFragment.createDisplayNameFromEmail(SignupEpilogueFragment.java:541)
     at org.wordpress.android.ui.accounts.signup.SignupEpilogueFragment.populateViews(SignupEpilogueFragment.java:622)
     at org.wordpress.android.ui.accounts.signup.SignupEpilogueFragment.onAccountChanged(SignupEpilogueFragment.java:478)
...
(4 additional frame(s) were not displayed)
```

### PR DESCRIPTION
**Pull request by planarvoid**
Fixes #12399 

This crash happens when you have an email like this `username.12.a@email.com`. The code that handles the username removes all the characters that are not a-z and we end up with an array of items like this `["username", "", "a"]`. As you see, the second item in the array is empty. This causes a crash because we're trying to call `substring(0, 1)` on this empty string. 

In this PR I've extracted the logic, added a check for empty string and wrote some tests. 

To test:
- Start with a fresh install of the app
- Click on "Sign up for WordPress.com"
- Select "Sign up with email"
- Use an email that contains letters having dots on both sides, for example for a Gmail account you can do something like this: `username+a.12.b@gmail.com` since the + gets stripped as well
- Go through the magic link signup
- Notice that the app no longer crashes, the suggested display name should be something like "Username A B"

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupUtils.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        for (item in usernameArray) {
+            if (item.isNotEmpty()) {
+                builder.append(item.substring(0, 1).toUpperCase(Locale.ROOT))
+                if (item.length > 1) {
+                    builder.append(item.substring(1))
+                }
+                builder.append(" ")
+            }
+        }
+        val displayName = builder.toString().trim { it <= ' ' }
```
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupUtils.kt:28]: Unless I'm missing something, I think `trim` can be simplified here:

```suggestion
        val displayName = builder.toString().trim()
```

File: WordPress/src/test/java/org/wordpress/android/ui/accounts/signup/SignupUtilsTest.kt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    private fun assertEmailMappedToDisplayName(email: String, expectedDisplayName: String? = null) {
+        val displayName = signupUtils.createDisplayNameFromEmail(email)
+        assertThat(displayName).isEqualTo(expectedDisplayName)
+    }
+
+    private fun assertEmailMappedToUsername(email: String, expectedDisplayName: String? = null) {
+        val displayName = signupUtils.createUsernameFromEmail(email)
+        assertThat(displayName).isEqualTo(expectedDisplayName)
+    }
```
Comment (renanferrari) [inline WordPress/src/test/java/org/wordpress/android/ui/accounts/signup/SignupUtilsTest.kt:91]: I believe those are some copy/paste typos? 
- `expectedDisplayName` → `expectedUsername`
- `displayName` → `username`

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupUtils.kt (hunk 1 above)
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupUtils.kt]: I think you're right. This is a result of the automatic conversion to kotlin

**Comment by planarvoid** [comment]
thanks for the review @renanferrari ! I've approved both of you suggested changes and I'm merging this PR

**Pull request merged**